### DISCUSSION TITLE
Select different length time slices and reduce with a function

### DISCUSSION THREAD

**Discussion by gbromley** [pydata/xarray#8103] [Q&A]
Hello! 

I have a data that looks like:

![Screenshot 2023-08-23 at 9 08 09 AM](https://github.com/pydata/xarray/assets/5695215/7970ad73-9d4e-42ab-a03d-412412611a77)

I would like to use the 'onset_date' and 'demise_date' together in a slice to select data (daily precipitation) from an xarray dataset and then apply a reducing function (such as mean). Onset_date and demise_date differ for each lat/lon/year combo. Selecting the time slices results in different length data arrays but then applying mean to them should output a single value for every lat/lon. 

Is there an efficient way to do this? I have to run this type of analysis on large datasets, so using apply_ufunc/dask would be ideal. Worst case scenario, I loop through every combination, but I'd like to avoid loops as best I can. 

Thanks!

**Comment by dcherian** [comment]
This sounds like a groupby problem assuming you can appropriately define the variable to group by

Can you write it the for loop version for a simple example?

**Reply by gbromley** [reply]
Sure, I'll get that written up today. I am working with some small test datasets right now, so a for loop should be fine for now. The analysis will extend to large earth system model simulations so figuring out a non-loop option might be good.

**Reply by dcherian** [reply]
I realised later that numpy's ufunc.reduceat fits what you want really well. Neither Xarray nor dask give you that at the moment but it should fit your core problem really well.
